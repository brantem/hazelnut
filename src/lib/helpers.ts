import days from 'data/days';
import { Day } from 'types/shared';

export const getMinutesFromTime = (time: string) => {
  const [hour, minute] = time.split(':');
  return parseInt(hour) * 60 + parseInt(minute);
};

export const getCurrentDay = (): Day => [...days.slice(6), ...days.slice(0, -1)][new Date().getDay()];
