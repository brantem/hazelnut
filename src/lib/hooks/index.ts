export * from 'lib/hooks/useSearch';
export * from 'lib/hooks/useModal';
