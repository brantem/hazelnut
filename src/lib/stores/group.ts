import create from 'zustand';

import { Group } from 'types/group';

type GroupState = {
  group: Group | null;
  clear: () => void;

  isSaveOpen: boolean;
  showSave: (group?: Group | null) => void;

  isSettingsOpen: boolean;
  showSettings: (group: Group) => void;

  hide: () => void;
};

export const useGroupStore = create<GroupState>()((set) => ({
  group: null,
  clear: () => set({ group: null }),

  isSaveOpen: false,
  showSave: (group = null) => set((state) => ({ group: group || state.group, isSaveOpen: true })),

  isSettingsOpen: false,
  showSettings: (group) => set({ group, isSettingsOpen: true }),

  hide: () => set({ isSaveOpen: false, isSettingsOpen: false }),
}));
