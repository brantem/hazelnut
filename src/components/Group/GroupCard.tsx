import { EllipsisHorizontalIcon } from '@heroicons/react/20/solid';
import { MinusCircleIcon } from '@heroicons/react/24/outline';
import clsx from 'clsx';

import DeleteButton from 'components/DeleteButton';

import { useGroupsStore, useItemsStore } from 'lib/stores';
import type { Group } from 'types/group';

type GroupCardProps = {
  group: Group;
};

const GroupCard = ({ group }: GroupCardProps) => {
  const { showSettings } = useGroupsStore();
  const items = useItemsStore((state) => state.items.filter((item) => item.groupId === group.id));
  const { showAdd, remove } = useItemsStore();

  return (
    <div className={`px-4 py-3 bg-${group.color}-50`} data-testid="group-card">
      <div className="flex items-center justify-between space-x-3">
        <h3 className={`text-sm font-semibold uppercase text-${group.color}-600 truncate`}>{group.title}</h3>

        <div className="flex flex-shrink-0 items-center space-x-1">
          <button
            className={`rounded-md px-2 py-1 text-sm hover:bg-${group.color}-100 flex-shrink-0`}
            onClick={() => showAdd(group.id)}
          >
            Add Item
          </button>

          <button
            className={`rounded-md p-1 hover:bg-${group.color}-100`}
            onClick={() => showSettings(group)}
            data-testid="group-card-settings"
          >
            <EllipsisHorizontalIcon className="h-5 w-5" />
          </button>
        </div>
      </div>

      {items.length ? (
        <ol className="space-y-1 pt-2 pb-1" data-testid="group-card-items">
          {items.map((item) => (
            <li
              data-testid="group-card-items-item"
              key={item.id}
              className="flex items-center justify-between space-x-3"
            >
              <span className="truncate">{item.title}</span>

              <DeleteButton
                className={(clicked) => clsx(`rounded-md p-1 text-sm text-red-500 hover:bg-red-100`, clicked && 'px-2')}
                text={<MinusCircleIcon className="h-5 w-5" />}
                onConfirm={() => remove(item.id)}
              />
            </li>
          ))}
        </ol>
      ) : null}
    </div>
  );
};

export default GroupCard;
