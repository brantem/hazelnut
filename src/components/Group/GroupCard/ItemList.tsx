import { useCallback } from 'react';
import { MinusCircleIcon } from '@heroicons/react/24/outline';
import clsx from 'clsx';

import DeleteButton from 'components/DeleteButton';

import { useItemsStore } from 'lib/stores';
import type { Group } from 'types/group';

type ItemListProps = {
  group: Group;
};

const ItemList = ({ group }: ItemListProps) => {
  const items = useItemsStore(useCallback((state) => state.getItemsByGroupId(group.id), [group.id]));
  const remove = useItemsStore((state) => state.remove);

  return (
    <ol className="space-y-1 pt-2 pb-1" data-testid="group-card-items">
      {items.map((item) => (
        <li data-testid="group-card-items-item" key={item.id} className="flex items-center justify-between space-x-3">
          <span className="truncate">{item.title}</span>

          <DeleteButton
            className={(clicked) => clsx(`rounded-md !p-1 text-sm text-red-500 hover:bg-red-100`, clicked && '!px-2')}
            text={<MinusCircleIcon className="h-5 w-5" />}
            onConfirm={() => remove(item.id)}
          />
        </li>
      ))}
    </ol>
  );
};

export default ItemList;
