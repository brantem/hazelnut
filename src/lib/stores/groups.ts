import { useState, useEffect } from 'react';
import create from 'zustand';
import { persist } from 'zustand/middleware';
import { nanoid } from 'nanoid';

import { Group, GroupItem } from 'types/group';

type GroupsState = {
  groups: Group[];
  add: (group: Omit<Group, 'id' | 'items'>) => void;
  addItem: (id: string, item: Omit<GroupItem, 'id'>) => void;
  edit: (id: string, group: Partial<Omit<Group, 'id' | 'items'>>) => void;
  remove: (id: string) => void;
  removeItem: (id: string, itemId: string) => void;
};

const useStore = create<GroupsState>()(
  persist(
    (set) => ({
      groups: [],
      add: (group) => {
        set((state) => ({ groups: [...state.groups, { id: nanoid(), items: [], ...group }] }));
      },
      addItem: (id, item) => {
        set((state) => ({
          groups: state.groups.map((group) =>
            group.id === id ? { ...group, items: [...group.items, { id: nanoid(), ...item }] } : group,
          ),
        }));
      },
      edit: (id, group) => {
        set((state) => ({
          groups: state.groups.map((_group) => (_group.id === id ? { ..._group, ...group } : _group)),
        }));
      },
      remove: (id) => {
        set((state) => ({ groups: state.groups.filter((group) => group.id !== id) }));
      },
      removeItem: (id, itemId) => {
        set((state) => ({
          groups: state.groups.map((group) =>
            group.id === id ? { ...group, items: group.items.filter((item) => item.id !== itemId) } : group,
          ),
        }));
      },
    }),
    { name: 'groups' },
  ),
);

/* c8 ignore start */
const dummy = {
  groups: [],
  add: () => {},
  addItem: () => {},
  edit: () => {},
  remove: () => {},
  removeItem: () => {},
};

// https://github.com/pmndrs/zustand/issues/1145
export const useGroupsStore = ((selector, equals) => {
  const store = useStore(selector, equals);
  const [isHydrated, setHydrated] = useState(false);
  useEffect(() => setHydrated(true), []);
  return isHydrated ? store : selector ? selector(dummy) : dummy;
}) as typeof useStore;
/* c8 ignore stop */
