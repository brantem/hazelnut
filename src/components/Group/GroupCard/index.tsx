import { useCallback, useMemo } from 'react';
import { EllipsisHorizontalIcon, PlusIcon, ChevronUpIcon } from '@heroicons/react/20/solid';
import clsx from 'clsx';

import ItemList from 'components/Group/GroupCard/ItemList';

import { isMatch } from 'lib/helpers';
import { useGroupsStore, useItemsStore, useSearchStore } from 'lib/stores';
import type { Group } from 'types/group';

type GroupCardProps = {
  group: Group;
};

const GroupCard = ({ group }: GroupCardProps) => {
  const { showAddItem, showSettings, edit } = useGroupsStore();
  const { search } = useSearchStore('items');
  const isGroupMatch = useMemo(() => {
    if (!search) return true;
    return isMatch(group.title, search);
  }, [group.title, search]);
  const isItemsMatch = useItemsStore(
    useCallback(
      (state) => {
        if (!search) return true;
        return state.getItemsByGroupId(group.id).findIndex((item) => isMatch(item.title, search)) !== -1;
      },
      [group.id, search],
    ),
  );

  if (!isGroupMatch && !isItemsMatch) return null;

  return (
    <div className={`px-4 py-3 bg-${group.color}-50`} data-testid="group-card">
      <div className="flex items-center justify-between space-x-3">
        <h3 className={`text-sm font-semibold uppercase text-${group.color}-600 truncate`}>{group.title}</h3>

        <div className="flex flex-shrink-0 items-center space-x-1">
          <button
            className={`rounded-md p-1 text-sm hover:bg-${group.color}-100 flex-shrink-0`}
            onClick={() => showAddItem(group)}
            data-testid="group-card-add-item"
          >
            <PlusIcon className="h-5 w-5" />
          </button>

          <button
            className={`rounded-md p-1 hover:bg-${group.color}-100`}
            onClick={() => showSettings(group)}
            data-testid="group-card-settings"
          >
            <EllipsisHorizontalIcon className="h-5 w-5" />
          </button>

          <button
            className={clsx(`rounded-md p-1 hover:bg-${group.color}-100`, group.minimized && 'rotate-180')}
            onClick={() => edit(group.id, { minimized: !group.minimized })}
            data-testid="group-card-minimize"
          >
            <ChevronUpIcon className="h-5 w-5" />
          </button>
        </div>
      </div>

      {!group.minimized && <ItemList group={group} />}
    </div>
  );
};

export default GroupCard;
