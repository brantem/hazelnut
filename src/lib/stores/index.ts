import { useEffect } from 'react';

import { itemsStore } from 'lib/stores/items';
import { groupsStore } from 'lib/stores/groups';
import { routinesStore } from 'lib/stores/routines';
import { historiesStore } from 'lib/stores/histories';
import storage from 'lib/stores/storage';

/* c8 ignore start */
export const useLoadStore = () => {
  useEffect(() => {
    (async () => {
      const [items, groups, routines, histories] = await Promise.all([
        storage.getAll('items'),
        storage.getAll('groups'),
        storage.getAll('routines'),
        storage.getAll('histories'),
      ]);

      itemsStore.setState({ items: items.sort((a, b) => a.createdAt - b.createdAt), isReady: true });
      groupsStore.setState({ groups: groups.sort((a, b) => a.createdAt - b.createdAt), isReady: true });
      routinesStore.setState({ routines, isReady: true });
      historiesStore.setState({ histories, selectedDate: localStorage.getItem('history-selected-date') });
    })();
  }, []);
};
/* c8 ignore stop */

export * from 'lib/stores/items';
export * from 'lib/stores/groups';
export * from 'lib/stores/routines';
export * from 'lib/stores/search';
export * from 'lib/stores/histories';
export * from 'lib/stores/modal';
