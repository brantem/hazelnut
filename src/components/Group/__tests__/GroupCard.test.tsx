import { render, screen, act, renderHook } from '@testing-library/react';
import '@testing-library/jest-dom';

import GroupCard from 'components/Group/GroupCard';

import { useGroupsStore, useItemsStore } from 'lib/stores';

describe('GroupCard', () => {
  beforeEach(() => {
    const items = renderHook(() => useItemsStore());
    act(() => items.result.current.add('group-1', { id: 'item-1', title: 'Item 1' } as any));
  });

  afterEach(() => {
    const items = renderHook(() => useItemsStore());
    act(() => items.result.current.remove('item-1'));
  });

  it('should render successfully', () => {
    const onAddItemClick = vi.fn(() => {});
    const onSettingsClick = vi.fn(() => {});
    const { container } = render(
      <GroupCard
        group={{ id: 'group-1', title: 'Group 1', color: 'red' }}
        onAddItemClick={onAddItemClick}
        onSettingsClick={onSettingsClick}
      />,
    );

    expect(container).toMatchSnapshot();
    expect(screen.getAllByTestId('group-card-items-item')).toHaveLength(1);

    act(() => screen.getByText('Add Item').click());
    expect(onAddItemClick).toHaveBeenCalled();

    act(() => screen.getByTestId('group-card-settings').click());
    expect(onSettingsClick).toHaveBeenCalled();
  });

  it('should render empty group', () => {
    const items = renderHook(() => useItemsStore());
    act(() => items.result.current.remove('item-1'));

    const { container } = render(
      <GroupCard
        group={{ id: 'group-1', title: 'Group 1', color: 'red' }}
        onAddItemClick={() => {}}
        onSettingsClick={() => {}}
      />,
    );

    expect(container).toMatchSnapshot();
    expect(screen.queryByTestId('group-card-items')).not.toBeInTheDocument();
  });

  it('should remove item', () => {
    const items = renderHook(() => useItemsStore());
    const remove = vi.spyOn(items.result.current, 'remove');

    const groups = renderHook(() => useGroupsStore());
    act(() => groups.result.current.add({ id: 'group-1', title: 'Group 1', color: 'red' } as any));

    const { rerender } = render(
      <GroupCard group={groups.result.current.groups[0]} onAddItemClick={() => {}} onSettingsClick={() => {}} />,
    );

    expect(screen.getByTestId('group-card-items-item')).toBeInTheDocument();
    act(() => screen.getByTestId('delete-button').click());
    act(() => screen.getByTestId('delete-button-confirm').click());
    expect(remove).toHaveBeenCalledWith('item-1');
    rerender(
      <GroupCard group={groups.result.current.groups[0]} onAddItemClick={() => {}} onSettingsClick={() => {}} />,
    );
    expect(screen.queryByTestId('group-card-items-item')).not.toBeInTheDocument();
  });
});
