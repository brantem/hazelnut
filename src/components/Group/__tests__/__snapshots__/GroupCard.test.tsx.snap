// Vitest Snapshot v1

exports[`GroupCard > should render empty group 1`] = `
<div>
  <div
    class="px-4 py-3 bg-red-50"
    data-testid="group-card"
  >
    <div
      class="flex justify-between items-center space-x-3"
    >
      <h3
        class="uppercase text-sm font-semibold text-red-600 truncate"
      >
        Group 1
      </h3>
      <div
        class="flex items-center space-x-1 flex-shrink-0"
      >
        <button
          class="px-2 py-1 text-sm rounded-md hover:bg-red-100 flex-shrink-0"
        >
          Add Item
        </button>
        <button
          class="p-1 rounded-md hover:bg-red-100"
          data-testid="group-card-settings"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 3a1.5 1.5 0 110 3 1.5 1.5 0 010-3zM10 8.5a1.5 1.5 0 110 3 1.5 1.5 0 010-3zM11.5 15.5a1.5 1.5 0 10-3 0 1.5 1.5 0 003 0z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`GroupCard > should render successfully 1`] = `
<div>
  <div
    class="px-4 py-3 bg-red-50"
    data-testid="group-card"
  >
    <div
      class="flex justify-between items-center space-x-3"
    >
      <h3
        class="uppercase text-sm font-semibold text-red-600 truncate"
      >
        Group 1
      </h3>
      <div
        class="flex items-center space-x-1 flex-shrink-0"
      >
        <button
          class="px-2 py-1 text-sm rounded-md hover:bg-red-100 flex-shrink-0"
        >
          Add Item
        </button>
        <button
          class="p-1 rounded-md hover:bg-red-100"
          data-testid="group-card-settings"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 3a1.5 1.5 0 110 3 1.5 1.5 0 010-3zM10 8.5a1.5 1.5 0 110 3 1.5 1.5 0 010-3zM11.5 15.5a1.5 1.5 0 10-3 0 1.5 1.5 0 003 0z"
            />
          </svg>
        </button>
      </div>
    </div>
    <ol
      class="space-y-1 pt-2 pb-1"
      data-testid="group-card-items"
    >
      <li
        class="flex justify-between items-center space-x-3"
        data-testid="group-card-items-item"
      >
        <span
          class="truncate"
        >
          Item 1
        </span>
        <button
          class="rounded-md hover:bg-red-100 text-red-500 text-sm p-1"
          data-testid="delete-button"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15 12H9m12 0a9 9 0 11-18 0 9 9 0 0118 0z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </li>
    </ol>
  </div>
</div>
`;
